anvill/src/Lifters/Context.cpp

```diff
@@ -69,71 +69,54 @@ llvm::Constant *EntityLifterImpl::TryLiftData(
   return nullptr;
 }
 
-// Tries to lift the entity at `address` and return an `llvm::Function *`
-// or `llvm::GlobalAlias *` relating to that address. The returned entity,
-// if any, will reside in `options.module`.
-llvm::Constant *EntityLifterImpl::TryLiftEntity(uint64_t address) {
-  auto ent_it = entities.find(address);
-  if (ent_it != entities.end()) {
-    return ent_it->second;
-  }
-
-  auto [byte, availability, permission] = memory_provider->Query(address);
-
-  switch (availability) {
-    case ByteAvailability::kUnknown:
-    case ByteAvailability::kAvailable:
-      break;
-
-    // If the byte isn't available, then it's not part of the address space
-    // to which the memory provider provides access.
-    case ByteAvailability::kUnavailable:
-      return nullptr;
-  }
-
-  llvm::Constant *ret = nullptr;
-
-  switch (permission) {
-    case BytePermission::kUnknown:
-    case BytePermission::kReadableExecutable:
-    case BytePermission::kReadableWritableExecutable:
-      if (auto maybe_decl = type_provider->TryGetFunctionType(address);
-          maybe_decl) {
-        ret = function_lifter.LiftFunction(*maybe_decl);
-        break;
+// Tells the entity lifter that `func` is the lifted function at `address`.
+// There is some collusion between the `EntityLifter` and the `FunctionLifter`
+// to ensure their view of the world remains consistent.
+void ContextImpl::SetLiftedFunction(uint64_t address, llvm::Function *func) {
+
+  address_to_entity[address] = func;
+  entity_to_address[func] = address;
+
+  // This function might call other functions, so make sure the entity lifter
+  // knows about the addresses of those other functions.
+  for (const auto &block : *func) {
```

Comment:
Maybe consider [llvm::instructions(const llvm::Function &)](https://llvm.org/doxygen/InstIterator_8h.html). Alternatively maybe an [llvm::InstVisitor](https://llvm.org/doxygen/classllvm_1_1InstVisitor.html) if you anticipate this code will handle more instruction types, beside calls.

Reply:
Oooh I was not aware of `llvm::instructions`!